frontend/device/android/device.lua

```diff
@@ -400,21 +408,38 @@ function Device:showLightDialog()
     local action = android.lights.dialogState()
     if action == C.ALIGHTS_DIALOG_OK then
         self.powerd.fl_intensity = self.powerd:frontlightIntensityHW()
+        if self.powerd.fl_intensity > self.powerd.fl_min then
+            self.powerd.is_fl_on = true
+        else
+            self.powerd.is_fl_on = false
+        end
         logger.dbg("Dialog OK, brightness: " .. self.powerd.fl_intensity)
         if android.isWarmthDevice() then
             self.powerd.fl_warmth = self.powerd:getWarmth()
             logger.dbg("Dialog OK, warmth: " .. self.powerd.fl_warmth)
         end
-        local Event = require("ui/event")
-        local UIManager = require("ui/uimanager")
-        UIManager:broadcastEvent(Event:new("FrontlightStateChanged"))
+        return true
     elseif action == C.ALIGHTS_DIALOG_CANCEL then
         logger.dbg("Dialog Cancel, brightness: " .. self.powerd.fl_intensity)
-        self.powerd:setIntensityHW(self.powerd.fl_intensity)
+        if not old_is_fl_sw_on then
+            local old_intensity = self.powerd.fl_intensity
+            self.powerd:setIntensityHW(self.powerd.fl_min)
+            self.powerd.fl_intensity = old_intensity
+            self.powerd.is_fl_on = false
+        else
+            if self.powerd.is_fl_on then
+                self.powerd:setIntensityHW(self.powerd.fl_intensity)
+            else
+                local old_intensity = self.powerd.fl_intensity
+                self.powerd:setIntensityHW(self.powerd.fl_min)
+                self.powerd.fl_intensity = old_intensity
+            end
+        end
         if android.isWarmthDevice() then
             logger.dbg("Dialog Cancel, warmth: " .. self.powerd.fl_warmth)
             self.powerd:setWarmth(self.powerd.fl_warmth)
         end
+        return false
```

Comment:
the whole function looks a bit complicate right now and I don't even understand why all that is needed :(

Reply:
Yes. Does not look simple. But with that HardwareSwitch things got complicated.
It could be, that the dialog is started with HardwareSwitch off of on.
We have to turn on HardwareSwitch on start of the dialog, so that someone can see the dialog in the dark (perhaps with the intensity which was before a toggle). 

At a cancel we have to restore the state before, but we have turned on the HardwareSwitch. If it was off before the dialog, we turn of is_fl_on and change the screen brightness to zero, but we also don't want to loose the old fl_intensity (`self.powerd:setIntensityHW(self.powerd.fl_min` changes fl_intensity). Here we don't need a broadcastLightChange.

At a accept we have to turn on is_fl_on, when the fl_intensity>fl_min, else if_sw_on==false.

Reply:
I still don't understand why we need to handle the state of the hw/sw switch. Benefits?. If I did understand correctly we use the switch to being able to change fl without root, but toggling the switch needs root, I'm totally lost here :(